sycl/include/sycl/detail/property_list_base.hpp

```diff
@@ -126,6 +127,17 @@ class PropertyListBase {
     }
   }
 
+  void
+  convertPropertiesToKinds(std::set<std::pair<int, bool>> &PropKinds) const {
+    for (size_t it = 0; it < MDataLessProps.size(); it++) {
+      if (MDataLessProps[it])
+        PropKinds.insert(std::pair{int(it), false});
+    }
+    for (auto &Prop : MPropsWithData) {
+      PropKinds.insert(std::pair{Prop->getKind(), true});
+    }
+  }
```

Comment:
If we move this to headers, we won't have to make sets of valid properties. We could instead just leverage the .def files to generate a switch-statement where default is false and the generated cases are true. Then we can send each of these kinds through the switches and base our decision on that.

One big benefit of doing it all in headers is that the property kinds will make the API volatile otherwise, as changing them could break code because the runtime library would keep the old kinds.

Reply:
I understand the idea of .def file usage (redefining __SYCL_DATA_LESS_PROP to convert that list to switch cases looks ok). 
It is still kind value based check but should work. I will try to make a different impl to check